### DISCUSSION TITLE
Immich constantly generating 85,000+ thumbnails

### DISCUSSION THREAD

**Discussion by 825i** [immich-app/immich#11043] [Q&A]
### The bug

I am trying to use Immich **v1.108.0** without machine learning (:noml) and using only an external library.
So no uploads through Immich.

Immich claims I have: 90496 photos and 1641 videos

It is constantly trying to generate 88,000+ thumbnails at an extremely slow pace. My full specs are below but the CPU is a `Ryzen 5 5600 6-Core @ 3500 MHz` which is moderately powerful. Thumbnail concurrency is set to only `10`.

What I don't understand is that the process has been running for days and the amount of thumbnails it said it needed to generate has gone from 82,000+ to now 88,000+ plus and higher. Whilst I do understand I have 90,000 photos, what I don't understand is that scrolling through my entire massive library. I already HAVE some kind of thumbnails. If I scroll up and down my library I can see already thumbnails of all my images. So what is this process doing, eating up literally 100% of my CPU over multiple days? It seems as though it's just infinitely generating thumbnails.

![Screenshot 2024-07-10 134610](https://github.com/immich-app/immich/assets/58630825/92800dba-f5f8-49e1-93d9-e717a8085b63)

Subfolder structure of the `/import` external library is like so with photos in year'd folders: 
![Screenshot 2024-07-11 193353](https://github.com/immich-app/immich/assets/58630825/f3842e80-d41f-4410-8a55-c31cbfd3c733)

After adding my external library at path: `/import`
which is mapped to: `/mnt/user0/storage/Personal/`

The path `/photos` is mapped to `/mnt/user0/storage/Immich/` which contains the following:

![image](https://github.com/immich-app/immich/assets/58630825/466db97d-b043-44f4-9524-cdd10cfeb262)

So this is in a different location. Nor does the external library folder also include the "thumbs" folder because I had read about that "infinite" thumbnail generation loop. So I avoided using the same DIR.

Machine specs:

```
Model: | Custom
M/B: | ASRock X570M Pro4 s/n M80-D9026600142
BIOS: | American Megatrends Inc. Version P3.90 Dated 10/19/2022
CPU: | AMD Ryzen 5 5600 6-Core @ 3500 MHz
HVM: | Enabled
IOMMU: | Disabled
Cache: | L1 - Cache: 384 KiB, L2 - Cache: 3 MiB, L3 - Cache: 32 MiB
Memory: | 32 GiB DDR4 Multi-bit ECC (max. installable capacity 128 GiB)
Network: | bond0: fault-tolerance (active-backup), mtu 9000
Kernel: | Linux 6.1.79-Unraid x86_64
OpenSSL: | 1.1.1v
Uptime: | 44 days, 5 hours, 55 minutes
```

### The OS that Immich Server is running on

Unraid 6.12.10

### Version of Immich Server

v1.108.0

### Version of Immich Mobile App

v1.107.2

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
Unraid does not use docker compose.
```

### Your .env content

```Shell
DB_PASSWORD=[REDACTED]
REDIS_PASSWORD=
PUID=99
MACHINE_LEARNING_HOST=0.0.0.0
HOSTNAME=45f5303454fd
S6_CMD_WAIT_FOR_SERVICES_MAXTIME=0
LANGUAGE=en_US.UTF-8
DB_HOSTNAME=192.168.1.50
HOST_CONTAINERNAME=immich
UMASK=022
PWD=/
DB_DATABASE_NAME=immich
NVIDIA_DRIVER_CAPABILITIES=compute,video,utility
NODE_ENV=production
TZ=[REDACTED]
DB_PORT=5433
HOME=/root
LANG=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:*~=00;90:*#=00;90:*.bak=00;90:*.crdownload=00;90:*.dpkg-dist=00;90:*.dpkg-new=00;90:*.dpkg-old=00;90:*.dpkg-tmp=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:*.swp=00;90:*.tmp=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:
PGID=100
VIRTUAL_ENV=/lsiopy
IMMICH_BUILD_DATA=/app/immich/server
IMMICH_MEDIA_LOCATION=/photos
S6_VERBOSITY=1
S6_STAGE2_HOOK=/docker-mods
IMMICH_PORT=8080
IMMICH_MACHINE_LEARNING_ENABLED=false
MACHINE_LEARNING_PORT=3003
DB_USERNAME=postgres
MACHINE_LEARNING_WORKERS=1
REDIS_HOSTNAME=192.168.1.50
TERM=xterm
HOST_OS=Unraid
SHLVL=1
MACHINE_LEARNING_WORKER_TIMEOUT=120
REDIS_PORT=6379
IMMICH_ENV=production
PATH=/lsiopy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOST_HOSTNAME=Tower
_=/usr/bin/env
```

### Reproduction steps

```bash
1. Set an external library
2. Thumbnail generation takes days
3. It seems thumbnails are being generated infinitely
4. Thumbnails already exist in the feed for literally all my images yet they are still generating
5. CPU usage stuck at 100% for days
```

### Relevant log output

```shell
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset b8281395-54ce-4de4-a562-947798aa9b34
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset 207adf15-b0ef-4dfb-972d-4710fb70fa32
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated JPEG image preview for asset d4c11b1e-3382-46ca-95a8-58420d75ab8b
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset bd357939-3dd6-4c56-a978-3fae5ed4c80a
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset a7bd9801-1661-48d2-9dc5-e07913525702
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset dcc0fe4d-26f6-4a56-9df6-ecd275c92996
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset 0baa20fd-ad41-4205-b48d-2fcd0c3192aa
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset fc981ec5-bbe9-4cfe-9fe9-4165dfa499e5
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset b2a432d6-8d8d-422f-b3a0-bc930363fe8a
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset 697c2193-f222-4341-a84f-f6e52bb3955c
[Nest] 53  - 07/11/2024, 7:25:29 PM     LOG [Microservices:MediaService] Successfully generated WEBP image thumbnail for asset d4c11b1e-3382-46ca-95a8-58420d75ab8b
```

**Comment by bo0tzz** [comment]
https://immich.app/docs/FAQ#why-are-there-so-many-thumbnail-generation-jobs

**Comment by 825i** [comment]
Yeah I have read this. The speed at which the thumbnail generation happens seems to be a problem. It's been multiple days and as I said, I seem to already have:

- Blurred (thumbhash)
- Preview (Webp)
- Thumbnail (Jpeg)

Thumbnails exist already for all my images when I actually go and look at them. So why is it still generating thumbnails for everything after a whole week? Is this really normal behaviour on a 6 core 3.5Ghz Ryzen 5 CPU?

If there are still 88,000 thumbnails to generate. Why can I see all of my images in the feed?

Lastly, and this to me is undeniable evidence that something is seriously wrong.

Why do I have a whopping 179,035 files and 49,800 folders in just my /thumbs/<usr-id> DIR?

![image](https://github.com/immich-app/immich/assets/58630825/bece2adc-ad92-4367-9a39-2f22e08a41a5)

**Comment by 825i** [comment]
@bo0tzz why did you close this without even waiting for a reply?

**Comment by 825i** [comment]
@alextran1502 could you kindly explain for posterity why after 7 days and 150,000 thumbnails and almost 50k folders in my `thumbs` directory. Why Immich is still telling me I have a further 88,000+ thumbnails to generate (the number has stayed the same roughly the last 7 days). I don't see how this should take so long. It seems like an instance of a previous bug that caused an infinite loop of generating thumbnails.

Bearing in mind that I have 90k images total and can already see all the blurred and regular thumbnails in my entire library when using Immich.

Could there at least be a clearer explanation for what thumbnails are being generated and even an ETA if this is seriously expected to take over 1 week on a 6-core 3.2Ghz CPU that has been running at 100% load now 24/7 for a week on a 10 concurrency limit and I was told that setting it higher was actually counterproductive.

**Comment by jrasm91** [comment]
Library scanning happens every day at midnight by default. I don't think it should be reprocessing files that are already imported, but if you turn this off does it finish processing all the jobs?

**Reply by 825i** [reply]
How do I turn the processing off? I can't see an option to disable the thumbnail generation at all and it isn't possible to set the jobs below 1. It tries to generate thumbnails any time I turn Immich on.

**Reply by jrasm91** [reply]
https://my.immich.app/admin/system-settings?isOpen=external-library+library-watching+library-scanning

You can also go to the jobs page and stop/clear all of the jobs:

https://my.immich.app/admin/jobs-status
